### PR TITLE
octopus: cephfs: backport snap-schedule module

### PR DESCRIPTION
**Pull request by jan--f**
Includes backports for #32319, #29489, #36877 and #36974 in this order.

and also #33888 (backport ticket: https://tracker.ceph.com/issues/47632)

Also added #37415  here.

And #34552, too :-)

And #36877 as well!

### CODE REVIEW

**Comment by smithfarm** [comment]
```
The following tests FAILED:
	  1 - run-tox-mgr (Failed)
```

**Comment by smithfarm** [comment]
`Docs: build check` fails with a 68 similar warnings:

```
WARNING: autodoc: failed to import method 'MgrModule.handle_command' from module 'mgr_module'; the following exception was raised:
/home/jenkins-build/build/workspace/ceph-pr-docs/build-doc/virtualenv/lib/python3.6/site-packages/cephfs.cpython-36m-x86_64-linux-gnu.so: undefined symbol: ceph_chdir
```

**Comment by smithfarm** [comment]
jenkins test make check

**Comment by jan--f** [comment]
jenkins test make check

**Comment by smithfarm** [comment]
@jan--f Does this need https://github.com/ceph/ceph/pull/36472 as well?

**Comment by jan--f** [comment]
> @jan--f Does this need #36472 as well?

I don't think so.

**Comment by smithfarm** [comment]
I looked into the `Docs: build check` failure in octopus backport PR https://github.com/ceph/ceph/pull/37142 and this message popped out at me:

```
+ ./admin/build-doc
Top Level States:  ['PeeringMachine']
Already using interpreter /usr/bin/python3
Using base prefix '/usr'
New python executable in /home/jenkins-build/build/workspace/ceph-pr-docs/build-doc/virtualenv/bin/python3
Also creating executable in /home/jenkins-build/build/workspace/ceph-pr-docs/build-doc/virtualenv/bin/python
Installing setuptools, pkg_resources, pip, wheel...done.
ERROR: After October 2020 you may experience errors when installing or updating packages. This is because pip will change the way that it resolves dependency conflicts.

We recommend you use --use-feature=2020-resolver to test your packages with the new resolver before it becomes the default.

sphinx-autodoc-typehints 1.11.0 requires Sphinx>=3.0, but you'll have sphinx 2.4.3 which is incompatible.
```

This seems suspicious, so I opened https://tracker.ceph.com/issues/47628 to track it.

@tchaikov Do you understand how "sphinx-autodoc-typehints 1.11.0" is getting installed in the Jenkins environment?

**Comment by smithfarm** [comment]
> @tchaikov Do you understand how "sphinx-autodoc-typehints 1.11.0" is getting installed in the Jenkins environment?

Ah, here it is. `admin/doc-requirements.txt`

```
Sphinx == 3.2.1
git+https://github.com/ceph/sphinx-ditaa.git@py3#egg=sphinx-ditaa
breathe
pyyaml >= 5.1.2
Cython
prettytable
sphinx-autodoc-typehints
sphinx-prompt
Sphinx-Substitution-Extensions
typed-ast
```

**Comment by smithfarm** [comment]
@jan--f Just for kicks, and only in the unlikely event you have nothing better to do, you could try cherry-picking #37389 into this PR just as an experiment to see if it fixes the `Docs: build check`?

**Comment by jan--f** [comment]
> @jan--f Just for kicks, and only in the unlikely event you have nothing better to do, you could try cherry-picking #37389 into this PR just as an experiment to see if it fixes the `Docs: build check`?

Sure no problem...done

**Comment by tchaikov** [comment]
might want to backport #33888 as well.

**Comment by jan--f** [comment]
@vshankar ready for review

**Comment by tchaikov** [comment]
@jan--f thanks for backporting #33888 , i 

- created https://tracker.ceph.com/issues/47632 and assigned it to you
- updated the first comment in your PR to point it to https://tracker.ceph.com/issues/47632

**Comment by smithfarm** [comment]
jenkins test api

**Comment by yuriw** [comment]
@jan--f pls rebase

**Comment by yuriw** [comment]
[wip-yuri8-testing-2020-10-05-1024-octopus](https://trello.com/c/k0zrv9PG/1034-wip-yuri8-testing-2020-10-05-1024-octopus)

**Comment by vshankar** [comment]
Let's get the master test PR (https://github.com/ceph/ceph/pull/34552) to merge before merging this.

DMN for now.

**Comment by smithfarm** [comment]
@jan--f Please rebase and, if possible, remind me why this is DNM?

**Comment by vshankar** [comment]
> @jan--f Please rebase and, if possible, remind me why this is DNM?

I marked it as DNM as the master test PR https://github.com/ceph/ceph/pull/34552 was pending merge.

**Comment by smithfarm** [comment]
> I marked it as DNM as the master test PR #34552 was pending merge.

OK, I trust you'll remove the DNM once that PR has been merged (?).

**Comment by vshankar** [comment]
> OK, I trust you'll remove the DNM once that PR has been merged (?).

absolutely -- once this rebased.

**Comment by jan--f** [comment]
ok, rebased and I cherry-picked #34552 commits as well. Ping @vshankar

**Comment by vshankar** [comment]
> ok, rebased and I cherry-picked #34552 commits as well. Ping @vshankar

thx

**Comment by smithfarm** [comment]
Flagging for integration test.

**Review by vshankar — APPROVED** [review]
LGTM

**Comment by yuriw** [comment]
@jan--f for some reason could not pull this PR, maybe needs rebase.

**Comment by smithfarm** [comment]
> @jan--f for some reason could not pull this PR, maybe needs rebase.

It conflicts with one or more of the other PRs in the integration branch. It won't need a rebase until those conflicting PRs are merged :-)

**Comment by smithfarm** [comment]
@jan--f Please rebase if you'd still like to see this merged.

**Comment by smithfarm** [comment]
@jan--f Moving this into integration testing (finally!)

File: src/pybind/mgr/mgr_module.py [hunk 1]
```diff
@@ -310,10 +310,21 @@ def __call__(self, func):
...
+        kwargs_switch = False
         for a, d in self.args_dict.items():
             if 'req' in d and d['req'] == "false" and a not in cmd_dict:
                 continue
-            kwargs[a.replace("-", "_")] = cmd_dict[a]
+            if kwargs_switch or (isinstance(cmd_dict[a], str) and '=' in cmd_dict[a]):
+                mgr.log.debug('found kwarg, assuming all following args are kw style')
+                kwargs_switch = True
+                try:
+                    k, arg = cmd_dict[a].split('=')
```
Comment (tchaikov) [inline src/pybind/mgr/mgr_module.py:321]:
```
2021-04-11T13:56:43.869 DEBUG:teuthology.orchestra.run.gibba003:> sudo /home/ubuntu/cephtest/cephadm --image quay.ceph.io/ceph-ci/ceph:33f0df555080d13dfd4cdf079a3f974d3b813ff4 shell -c /etc/ceph/ceph.conf -k /etc/ceph/ceph.client.admin.keyring --fsid 6b93e0fc-9acd-11eb-81
a3-001a4aab830c -- ceph orch daemon add mon 'gibba003:[v2:172.21.2.103:3301,v1:172.21.2.103:6790]=c'
2021-04-11T13:56:45.795 INFO:teuthology.orchestra.run.gibba003.stderr:Error EINVAL: Traceback (most recent call last):
2021-04-11T13:56:45.796 INFO:teuthology.orchestra.run.gibba003.stderr:  File "/usr/share/ceph/mgr/mgr_module.py", line 1223, in _handle_command
2021-04-11T13:56:45.796 INFO:teuthology.orchestra.run.gibba003.stderr:    return self.handle_command(inbuf, cmd)
2021-04-11T13:56:45.796 INFO:teuthology.orchestra.run.gibba003.stderr:  File "/usr/share/ceph/mgr/orchestrator/_interface.py", line 140, in handle_command
2021-04-11T13:56:45.796 INFO:teuthology.orchestra.run.gibba003.stderr:    return dispatch[cmd['prefix']].call(self, cmd, inbuf)
2021-04-11T13:56:45.797 INFO:teuthology.orchestra.run.gibba003.stderr:  File "/usr/share/ceph/mgr/mgr_module.py", line 331, in call
2021-04-11T13:56:45.797 INFO:teuthology.orchestra.run.gibba003.stderr:    return self.func(mgr, **kwargs)
2021-04-11T13:56:45.797 INFO:teuthology.orchestra.run.gibba003.stderr:  File "/usr/share/ceph/mgr/orchestrator/_interface.py", line 102, in <lambda>
2021-04-11T13:56:45.797 INFO:teuthology.orchestra.run.gibba003.stderr:    wrapper_copy = lambda *l_args, **l_kwargs: wrapper(*l_args, **l_kwargs)
2021-04-11T13:56:45.798 INFO:teuthology.orchestra.run.gibba003.stderr:  File "/usr/share/ceph/mgr/orchestrator/_interface.py", line 91, in wrapper
2021-04-11T13:56:45.798 INFO:teuthology.orchestra.run.gibba003.stderr:    return func(*args, **kwargs)
2021-04-11T13:56:45.798 INFO:teuthology.orchestra.run.gibba003.stderr:TypeError: _daemon_add_misc() got an unexpected keyword argument 'gibba003:[v2:172.21.2.103:3301,v1:172.21.2.103:6790]'
2021-04-11T13:56:45.798 INFO:teuthology.orchestra.run.gibba003.stderr:
2021-04-11T13:56:46.267 DEBUG:teuthology.orchestra.run:got remote process result: 22
```

i think this function somehow took `'gibba003:[v2:172.21.2.103:3301,v1:172.21.2.103:6790]=c'` as a kwarg, and used `"gibba003:[v2:172.21.2.103:3301,v1:172.21.2.103:6790]"` as the key.

see https://pulpito.ceph.com/kchai-2021-04-11_12:37:49-rados-wip-kefu-octopus-testing-2021-04-11-1851-distro-basic-gibba/6036056/

probably we should backport 922f962ae123b94ee7f92722190d19280f4ef6e3 as well.

**Review by tchaikov — CHANGES_REQUESTED** [review]

File: src/pybind/mgr/mgr_module.py (hunk 1 above)
Comment (jan--f) [reply src/pybind/mgr/mgr_module.py]: added

**Comment by batrick** [comment]
many test failures in: https://pulpito.ceph.com/yuriw-2021-04-21_21:21:36-fs-wip-yuri6-testing-2021-04-21-0941-octopus-distro-basic-smithi/

**Comment by jan--f** [comment]
I was sure but not longer am :smile: lets close it.

**Pull request closed without merging**